src/vm/comutilnative.cpp

```diff
@@ -2624,14 +2623,173 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+            {
+                canCompareBitsOrUseFastGetHashCode = FALSE;
+                break;
+            }
+        }
+        else if (pField->GetFieldType() == ELEMENT_TYPE_R8
+                || pField->GetFieldType() == ELEMENT_TYPE_R4)
+        {
+            // We have double/single field, cannot compare in fast path.
+            canCompareBitsOrUseFastGetHashCode = FALSE;
+            break;
+        }
+    }
+
+    // We've gone through all instance fields. It's time to cache the result.
+    // Note SetCanCompareBitsOrUseFastGetHashCode(BOOL) ensures the checked flag
+    // and canCompare flag being set atomically to avoid race.
+    mt->SetCanCompareBitsOrUseFastGetHashCode(canCompareBitsOrUseFastGetHashCode);
+
+    return canCompareBitsOrUseFastGetHashCode;
+}
+
+static BOOL CanUseFastGetHashCodeHelper(MethodTable *mt)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+    } CONTRACTL_END;
+
+    if (mt->ContainsPointers()
+        || mt->IsNotTightlyPacked())
+    {
+        mt->SetHasCheckedCanCompareBitsOrUseFastGetHashCode();
+        return FALSE;
+    }
+
+    return CanCompareBitsOrUseFastGetHashCodeWorker(mt, 
+                                                   MscorlibBinder::GetClass(CLASS__VALUE_TYPE),
+                                                   MscorlibBinder::GetMethod(METHOD__VALUE_TYPE__EQUALS)->GetSlot(),
+                                                   MscorlibBinder::GetMethod(METHOD__VALUE_TYPE__GET_HASH_CODE)->GetSlot());;
+}
+
+NOINLINE static BOOL CanCompareBitsHelper(MethodTable* mt, OBJECTREF objRef)
+{
+    FC_INNER_PROLOG(ValueTypeHelper::CanCompareBits);
+
+    _ASSERTE(mt != NULL);
+    _ASSERTE(objRef != NULL);
+
+    BOOL ret = FALSE;
+
+    HELPER_METHOD_FRAME_BEGIN_RET_ATTRIB_1(Frame::FRAME_ATTR_EXACT_DEPTH|Frame::FRAME_ATTR_CAPTURE_DEPTH_2, objRef);
+
+
+    ret = CanCompareBitsOrUseFastGetHashCodeWorker(mt, 
+                                                   MscorlibBinder::GetClass(CLASS__VALUE_TYPE),
+                                                   MscorlibBinder::GetMethod(METHOD__VALUE_TYPE__EQUALS)->GetSlot(),
+                                                   MscorlibBinder::GetMethod(METHOD__VALUE_TYPE__GET_HASH_CODE)->GetSlot());
+
+    HELPER_METHOD_FRAME_END();
+    FC_INNER_EPILOG();
+    
+    return ret;
+}
+
+// Return true if the valuetype does not contain pointer, is tightly packed, 
+// does not have floating point number field and does not override Equals method.
 FCIMPL1(FC_BOOL_RET, ValueTypeHelper::CanCompareBits, Object* obj)
 {
     FCALL_CONTRACT;
 
     _ASSERTE(obj != NULL);
     MethodTable* mt = obj->GetMethodTable();
-    FC_RETURN_BOOL(!mt->ContainsPointers() && !mt->IsNotTightlyPacked());
+    if (mt->ContainsPointers()
+        || mt->IsNotTightlyPacked())
+    {
+        mt->SetHasCheckedCanCompareBitsOrUseFastGetHashCode();
```

Comment:
Move `ContainsPointers` and `IsNotTightlyPacked` into `CanCompareBitsOrUseFastGetHashCodeWorker` would probably slow down the type that do has pointers or not tightly packed. How about move the `ContainsPointers` and `IsNotTightlyPacked` after following check?:

```cpp
 if (mt->HasCheckedCanCompareBitsOrUseFastGetHashCode())
    {
        FC_RETURN_BOOL(mt->CanCompareBitsOrUseFastGetHashCode());
    }
```